### PR TITLE
doc/options: Fix sample shellpipe PowerShell config.

### LINKED ISSUE #13713: Docs for shell-powershell suggest setting shellpipe=\|, but shellpipe is actually used to redirect to a file
The documentation under [`:help shell-powershell`](https://neovim.io/doc/user/options.html#shell-powershell) suggests using this configuration:

<pre>
let &shell = has('win32') ? 'powershell' : 'pwsh'
set shellquote= <b>shellpipe=\|</b> shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=\|\ Out-File\ -Encoding\ UTF8
</pre>

(Docs for powershell were originally introduced in f72f638f976347e2f14fc6ae5e5293c02d5ccfc1.)

But this setting of `'shellpipe'` is inconsistent with how `:make` uses it. It wants to use `'shellpipe'` followed by a file name to redirect output to that file. (So, yeah, the setting is definitely a misnomer!)

It would probably be more appropriate to use:

<pre>
set shellpipe=\|\ Out-File\ -Encoding\ UTF8
</pre>

(But then, I'm not a Windows or powershell guy, so I wouldn't know for sure...)

Using a simple `>` for this redirect doesn't seem to work either... It seems to produce a UTF-16 file and NeoVim then has trouble with the Unicode BOM at the start of it.

This issue was uncovered in the investigation of this Vi & Vim Stack Exchange question: [":make appending strange folder to my makeprg on Windows"](https://vi.stackexchange.com/q/28676/18609). The issue with `shellpipe=>` (which is its default value on Windows) was then uncovered in the follow up question [":make not echoing output to the command line"](https://vi.stackexchange.com/q/28684/18609).

### PR DESCRIPTION
**Pull request by dtcaciuc**
Capture error stream that usually contains compile error trace
and use Out-File to ensure output is written out as UTF8

Closes #13713

### CODE REVIEW

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -5205,11 +5205,12 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 	unescaping, so to keep yourself sane use |:let-&| like shown above.
 							*shell-powershell*
-	To use powershell: >
+	To use PowerShell: >
 		let &shell = has('win32') ? 'powershell' : 'pwsh'
-		set shellquote= shellpipe=\| shellxquote=
+		set shellquote= shellxquote=
 		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
 		set shellredir=\|\ Out-File\ -Encoding\ UTF8
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8
```
Comment (janlazo) [inline runtime/doc/options.txt:5213]: Why is `2>&1` prefixed on `shellpipe` only?

**Review by janlazo — CHANGES_REQUESTED** [review]
Needs a test.

`set_shell_powershell` in `test/functional/helpers.lua` should be updated.

**Comment by dtcaciuc** [comment]
Yes, I left it in WIP and will try to get tests running on my machine.

Speaking of tests, seems like pipelines occasionally fail to download lua rocks. I tried one failed URL locally and it worked. Do we know why this is happening?

File: runtime/doc/options.txt (hunk 1 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5213]: I omitted it because I was under impression that, for filtering commands, we should not be including error text together with the filtered data that comes out of standard output. However, reading the manual again, it says 
```
For Win32, the Unix checks are done and additionally "cmd" is checked for, which makes the default ">%s 2>&1
```
So, to be consistent, I will change that to capture error stream as well.

I also noticed that, even with `shellredir` configured, I still can't run stuff like `!Select-String` because Vim turns it into `SelectString < filename`. Looks like that pattern is [hardcoded](https://github.com/neovim/neovim/blob/ce976bf3559edca6d3a649c9053fab88959de6cf/src/nvim/ex_cmds.c#L1541), but I don't think it works with PowerShell.

File: test/functional/ex_cmds/make_spec.lua [hunk 2]
```diff
@@ -2,25 +2,30 @@ local helpers = require('test.functional.helpers')(after_each)
...
+local has_powershell = helpers.has_powershell
 local nvim = helpers.nvim
 local nvim_dir = helpers.nvim_dir
 
 describe(':make', function()
 
   describe('with powershell', function()
+      before_each(function ()
+        clear()
+        if has_powershell() then
```
Comment (janlazo) [inline test/functional/ex_cmds/make_spec.lua:14]: Can you skip the entire test suite instead if the user doesn't have powershell?

Comment (dtcaciuc) [reply test/functional/ex_cmds/make_spec.lua:14]: Sounds good, i will try that. I was going by the README instruction which states
```
skip the test via pending() if the external factor is missing. This ensures that the total test-count (success + fail + error + pending) is the same in all environments.
```

File: runtime/doc/options.txt (hunk 1 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5213]: I found another potential issue where both `powershell` and `pwsh` seem to have a couple of behavioral quirks:

Both `powershell` and `pwsh` considers a binary to have failed if it writes anything to stderr (it appears that `$?=True` is a "success" flag and is equivalent to Unix's `$?=0`)
```
# Writes to stdout; returns 0
PS C:\Users\dtcaciuc\code\neovim\build> .\bin\shell-test.exe REP 3 hello 2>&1; $?
0: hello
1: hello
2: hello
True
# Writes to stderr; returns 0
PS C:\Users\dtcaciuc\code\neovim\build> .\bin\shell-test.exe 2>&1; $?
ready $
False
```
This results in nvim printing `shell returned 1`

Additionall, `powershell` also writes out an exception object alongside with the program output an so running `r !bin/shell-test` results in the following, which is not ideal:
```
bin\shell-test.exe : ready $
At line:1 char:1
+ bin\shell-test.exe 2>&1 | Out-File -Encoding UTF8 C:\Users\dtcaciuc\A ...
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (ready $ :String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```

I'm still going through docs to try to see if there's a way to reconcile these.

Comment (janlazo) [reply runtime/doc/options.txt:5213]: Check `$LASTEXITCODE` for binaries, `$?` for powershell code.

Comment (dtcaciuc) [reply runtime/doc/options.txt:5213]: Ok, so after digging in this some more, my take on this is that Powershell 5 that comes bundled with Windows has two problematic behaviors
* Redirect commands seem to write data in odd encoding as reported in the issue
* Whenever caller redirects error stream with 2> and if the executed binary writes anything to error stream, it always turns into an error object (even if the content is a warning), which in turn results in an ugly error trace that I'm not able to silence. I can set ` $ErrorActionPreference = 'SilentlyContinue'` but then it completely discards the error  stream output together with the error object.

Unless there's some trick I could not find, I do not think that it's possible to come up with a good config. The error object behaviour be seen in this simple example and I can't figure out how to work around it.

```
PS C:\Users\dtcaciuc\code\neovim\build> powershell -Command '.\bin\shell-test.exe FOO >out.txt 2>&1'
PS C:\Users\dtcaciuc\code\neovim\build> cat .\out.txt
.\bin\shell-test.exe : Unknown first argument: FOO
At line:1 char:1
+ .\bin\shell-test.exe FOO >out.txt 2>&1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Unknown first argument: FOO:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

```

On the other hand, PowerShell Core (in my case, 7.1.3) seems to behave as expected by default. It produces encoding that Vim understands as-is and it also doesn't seem to automatically turn any error stream content into an error object. Now, the confusing part about this is from what I can gather, it still SHOULD. There's an [experimental feature](https://docs.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.1#psnotapplyerroractiontostderr) that supposedly needs to be enabled explicitly , but in my case it's off, so I'm not sure what is happening here. Here's an illustration 

```
PS C:\Users\dtcaciuc\code\neovim\build> pwsh -Command 'Get-ExperimentalFeature PSNotApplyErrorActionToStderr'

Name                                Enabled Source                              Description
----                                ------- ------                              -----------
PSNotApplyErrorActionToStderr         False PSEngine                            Don't have $ErrorActionPreference affect stderr …

PS C:\Users\dtcaciuc\code\neovim\build> pwsh -Command '.\bin\shell-test.exe FOO >out.txt 2>&1'
PS C:\Users\dtcaciuc\code\neovim\build> cat .\out.txt
Unknown first argument: FOO
```


Overall, the following is the configuration that works for both PS and PS Core. For PS, this does not tee make output to command window and it produces ugly error traces, but it's functional. For PS Core, it improves on PS behaviour by not printing error objects to output.
 
```
" PowerShell
let &shell = 'powershell'
set shellquote= shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
```

Furthermore, here's a configuration that works even better on PS Core. It causes `:make` and `:r  !<executable>` commands behave identically in my Windows and WSL environments.

```
" PowerShell Core
let shell = 'pwsh'
set shellquote= shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=2>&1\ \|\ Out-File\ %s;\ exit\ $LastExitCode
set shellpipe=2>&1\ \|\ Tee-Object\ %s;\ exit\ $LastExitCode
```

I intend to modify the test config and help doc to include the first PowerShell config because I assume that's what's available on base Windows machines and test hosts. Please let me know if I should add the additional PowerShell Core config to help document as well.

File: test/functional/ex_cmds/make_spec.lua (hunk 2 above)
Comment (dtcaciuc) [reply test/functional/ex_cmds/make_spec.lua]: Done; PTAL

File: runtime/doc/options.txt (hunk 1 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: Added a blurb about PowerShell 7.1+

Looks like this is the change that is causing the difference https://github.com/PowerShell/PowerShell/pull/13395

**Comment by dtcaciuc** [comment]
Finally found how my change was breaking the tests; it took way too many commits to get there, sorry about that! This should be good to go now.

File: runtime/doc/options.txt [hunk 3]
```diff
@@ -5205,11 +5205,25 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8;\ exit\ $LastExitCode
+
+< 	The above should work for all versions of PowerShell, but may result
+	in superfluous error object traces with PowerShell 5.1 due to how it
+	reacts when a running program writes anything to standard error stream.
+	This seems to be fixed in PowerShell versions 7.1 and above. The following
+	config may also be used to get a more Unix-like behaviour and echo make
+	output in the command window: >
+
+		let &shell = 'pwsh'
```
Comment (janlazo) [inline runtime/doc/options.txt:5222]: Can you create a test for `pwsh` and skip it if it is unavailable?

File: runtime/doc/options.txt [hunk 4]
```diff
@@ -5205,11 +5205,25 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+		set shellquote= shellxquote=
 		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
-		set shellredir=\|\ Out-File\ -Encoding\ UTF8
+		set shellredir=2>&1\ \|\ Out-File\ -Encoding\ UTF8;\ exit\ $LastExitCode
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8;\ exit\ $LastExitCode
+
+< 	The above should work for all versions of PowerShell, but may result
+	in superfluous error object traces with PowerShell 5.1 due to how it
+	reacts when a running program writes anything to standard error stream.
+	This seems to be fixed in PowerShell versions 7.1 and above. The following
```
Comment (janlazo) [inline runtime/doc/options.txt:5218]: Need a link to the release or changelog that mentions this fix.
Add it in the commit message, not this file, when you're ready to rebase.

File: runtime/doc/options.txt (hunk 3 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5222]: PTAL; I've split up pwsh bits into separate has_pwsh()/set_pwsh_shell()

File: runtime/doc/options.txt (hunk 4 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5218]: If I understand you right, does this mean that I should collapse all commits into one when I do the rebase?

**Comment by dtcaciuc** [comment]
Somewhat unrelated, but one of the channel tests seems flaky. Second to last pipeline gave out this

```
2021-04-18T02:10:50.9193044Z   [----------] Running tests from test/functional\core\channels_spec.lua
2021-04-18T02:10:50.9194420Z   [ RUN      ] channels can connect to socket: 0.00 ms SKIP
2021-04-18T02:10:51.0145663Z   [ RUN      ] channels can use stdio channel: 63.00 ms ERR
2021-04-18T02:10:51.0147429Z   test\helpers.lua:73: Expected objects to be the same.
2021-04-18T02:10:51.0147905Z   Passed in:
2021-04-18T02:10:51.0148222Z   (table: 0x027c775e26b8) {
2021-04-18T02:10:51.0148576Z     [1] = 'notification'
2021-04-18T02:10:51.0148904Z    *[2] = 'stdout'
2021-04-18T02:10:51.0149159Z     [3] = {
2021-04-18T02:10:51.0149413Z       [1] = 3
2021-04-18T02:10:51.0149648Z       [2] = {
2021-04-18T02:10:51.0149910Z         [1] = '' } } }
2021-04-18T02:10:51.0150184Z   Expected:
2021-04-18T02:10:51.0150502Z   (table: 0x027c76296a80) {
2021-04-18T02:10:51.0150830Z     [1] = 'notification'
2021-04-18T02:10:51.0151190Z    *[2] = 'stderr'
2021-04-18T02:10:51.0151447Z     [3] = {
2021-04-18T02:10:51.0151696Z       [1] = 3
2021-04-18T02:10:51.0151933Z       [2] = {
2021-04-18T02:10:51.0152199Z         [1] = '' } } }
```

Looks like it's... decided to write to stdout instead of stderr? I re-triggered the pipeline and the error went away. Should I file a bug issue?

File: test/functional/helpers.lua [hunk 5]
```diff
@@ -508,7 +508,9 @@ end
 
 -- PowerShell 5.1; comes with and available only on Windows
 function module.has_powershell()
-  return module.eval('executable("powershell")') == 1
+  -- Ensure we're on Windows in case, for whatever reason, pwsh is
+  -- aliased to powershell like it apparently is on macos test runners.
```
Comment (janlazo) [inline test/functional/helpers.lua:512]: Come up with a different os-agnostic check for `has_powershell` (and `has_pwsh`) to tolerate aliases or revert the change in `has_powershell`.

Comment (dtcaciuc) [reply test/functional/helpers.lua:512]: Sorry, I don't understand what you mean. Originally, both functions literally had the same `iswin()` check in them, I did not change anything there. The only thing that changed is has_powershell() guarded tests now only run on Windows.

Test suite assumed `pwsh` and `powershell` commands can be used interchangeably, and they can't (at least not at all times). `powershell` is PowerShell 5 and exists only on Windows and `pwsh` is PowerShell Core which exhibits different behaviour and allow us to use a different recommended configuration, which sometimes produces a different output, and thus needs a different set of test reference values.

So, test writers are basically dealing with two different shells here that may need to be tested separately. 

The split also allows us to now test both PowerShell 5 and PowerShell Core on Windows runners (previously, only PowerShell 5 was tested)

Comment (dtcaciuc) [reply test/functional/helpers.lua]: PTAL. I'm not sure I understood you correctly, but I reverted the change and right now we have

* has_powershell() -- tries windows powershell 5 on windows and falls back to powershell core everywhere else
* has_pwsh() -- specifically checks for powershell core

File: runtime/doc/options.txt [hunk 6]
```diff
@@ -5205,11 +5205,25 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 		let &shell = has('win32') ? 'powershell' : 'pwsh'
-		set shellquote= shellpipe=\| shellxquote=
+		set shellquote= shellxquote=
 		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
-		set shellredir=\|\ Out-File\ -Encoding\ UTF8
+		set shellredir=2>&1\ \|\ Out-File\ -Encoding\ UTF8;\ exit\ $LastExitCode
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8;\ exit\ $LastExitCode
+
+< 	The above should work for all versions of PowerShell, but may result
+	in superfluous error object traces with PowerShell 5.1 due to how it
```
Comment (janlazo) [inline runtime/doc/options.txt:5216]: Is powershell v5 what you have on your system? Windows 7 should be supported because of MSVC even if the OS is EOL. This means Powershell v2-5 should be considered if you're going to mention exact version numbers.

I suggest rewriting the paragraph to target Powershell v7.1 and suggest the above config for older powershell versions unless there's powershell v6 issue like with php.

File: runtime/doc/options.txt (hunk 4 above)
Comment (janlazo) [reply runtime/doc/options.txt:5218]: Make your commits atomic such that each commit has a single purpose and passes the build.

Squashing everything to 1 commit is fine as long as you can describe every change in 1 commit message.

File: test/functional/helpers.lua [hunk 7]
```diff
@@ -518,8 +519,29 @@ function module.set_shell_powershell()
...
+function module.set_shell_pwsh()
+  -- NOTE This config works with current PowerShell Core 7.1.3
+  assert(module.has_pwsh())
+  local cmd = 'Remove-Item -Force '..table.concat(iswin()
+    and {'alias:cat', 'alias:echo', 'alias:sleep'}
+    or  {'alias:echo'}, ',')..';'
+  module.source([[
+    let &shell = 'pwsh'
+    set shellquote= shellxquote=
+    let &shellredir='2>&1 | Out-File %s; exit $LastExitCode'
```
Comment (janlazo) [inline test/functional/helpers.lua:543]: Always UTF8, even on Windows?

**Comment by janlazo** [comment]
Ignore the flaky functional tests. LSP, channels, jobs, and any async test are flaky in Windows and Mac. BSD may error out due to some (screen,LSP) bug but it's likely unrelated to powershell.

File: runtime/doc/options.txt (hunk 1 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: I started checking different powershell versions and it looks like v6 doesn't have the error object problem either... Strangely enough, powershell 2 on windows 7 also doesn't seem to have it, only powershell 5 on windows 10 does. Powershell is a pretty frustrating program.

Either way, the 7.1 change above is irrelevant if we use `exit $LastExitCode`, so I don't think there's any point in mentioning it.

File: test/functional/helpers.lua (hunk 7 above)
Comment (dtcaciuc) [reply test/functional/helpers.lua:543]: According to [this doc](https://docs.microsoft.com/en-us/powershell/scripting/dev-cross-plat/vscode/understanding-file-encoding?view=powershell-7.1#configuring-powershell), yes. 

> In PowerShell 6+, the default encoding is UTF-8 without BOM on all platforms.

File: runtime/doc/options.txt (hunk 6 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: Rewrote for clarity; PTAL

File: test/functional/helpers.lua (hunk 7 above)
Comment (dtcaciuc) [reply test/functional/helpers.lua:543]: But naturally, what's actually happening is not following the docs 😑... Back to the drawing board
```
PowerShell 7.1.3
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

PS C:\Users\dtcaciuc> [Console]::OutputEncoding

IsSingleByte      : True
EncodingName      : OEM United States
WebName           : ibm437
HeaderName        : ibm437
BodyName          : ibm437
```

**Comment by dtcaciuc** [comment]
So it turns out if Tee-Object receives no input, it doesn't write out any file at all... which doesn't work for nvim since it expects an error file from `:make` whether it was successful or not.

Thus, powershell and pwsh will end up having the same config. Do you still want separate test case for pwsh or can I blow it away and just leave has_powershell()/.set_shell_powershell()?

**Comment by janlazo** [comment]
Sure, you can blow it away if it doesn't serve a technical use for Neovim at this moment.

**Comment by dtcaciuc** [comment]
Alright, everything is cleaned up and rebased; PTAL

File: runtime/doc/options.txt [hunk 8]
```diff
@@ -5227,11 +5227,12 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 	unescaping, so to keep yourself sane use |:let-&| like shown above.
 							*shell-powershell*
-	To use powershell: >
+	To use PowerShell: >
 		let &shell = has('win32') ? 'powershell' : 'pwsh'
-		set shellquote= shellpipe=\| shellxquote=
-		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
-		set shellredir=\|\ Out-File\ -Encoding\ UTF8
+		set shellquote= shellxquote=
+		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command\ [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;
```
Comment (janlazo) [inline runtime/doc/options.txt:5233]: Need a test to justify putting this in the docs or reference because the option value is so long.. Existing tests didn't need to change default input/output encoding of powershell.

File: runtime/doc/options.txt [hunk 9]
```diff
@@ -5227,11 +5227,12 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+		set shellredir=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
```
Comment (janlazo) [inline runtime/doc/options.txt:5235]: Use `let &option` syntax. Too many escaped characters.

File: runtime/doc/options.txt (hunk 8 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5233]: I already added this to set_shell_powershell() so it should be tested everywhere that helper is used, no?

I came up with this config by going through all of PS2, 5, and 7 and this is the most consistent one that worked for all of those. Something that looked like a one line doc fix turned out to be a really difficult problem heh.

File: runtime/doc/options.txt (hunk 9 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5235]: Ok, I simply followed what was there before. I'll change it to use let.

File: runtime/doc/options.txt (hunk 8 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt:5233]: I think I'll need help here. I tried writing something involving unicode chars before while working on this but I could not find a correct way to enter either input or reference values or both. There's also a few test harness styles as far as I can see (screen content tests, then the ones that seem to be getting buffer contents directly, etc), which one I should be using? Is there one that's close to what I would be writing here?

**Comment by janlazo** [comment]
Vim added powershell support in https://github.com/vim/vim/commit/127950241e84c822d3c50f46a00d42a70d2d5cb6 .

**Comment by dtcaciuc** [comment]
@janlazo Alright, so then... what's the plan? Do I close this PR then and you guys will just apply that patch instead?

**Comment by janlazo** [comment]
@dtcaciuc The plan is to merge this and improve that patch with our shell options because they didn't handle pwsh.

File: runtime/doc/options.txt (hunk 8 above)
Comment (janlazo) [reply runtime/doc/options.txt:5233]: Screen test is fine for validating powershell output in the buffer.

I'd start with `system()`, `systemlist`, `jobstart()`) which can accept stdin for unicode characters and can pass its output to Vimscript  variables so that you don't need to use screen tests.
The tests should set the shell options directly to assert the expected results on different input/output encodings.

**Comment by dtcaciuc** [comment]
Rebased the branch to latest master and bumped into https://github.com/neovim/neovim/issues/8472

File: runtime/doc/options.txt (hunk 8 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: @janlazo PTAL; see new test in eval/system_spec.lua. I looked at other tests and some of them do seem to verify default vim config (such as utf8 encoding), so first part my test uses and verifies the helper config, then ensures that config actually does something.

File: runtime/doc/options.txt (hunk 9 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: Done; PTAL

File: runtime/doc/options.txt (hunk 8 above)
Comment (janlazo) [reply runtime/doc/options.txt]: That's fine. Create a similar test for `systemlist()` and I'll merge.

File: test/functional/eval/system_spec.lua [hunk 10]
```diff
@@ -174,6 +174,19 @@ describe('system()', function()
...
+      if not helpers.has_powershell() then
+        pending("not tested; powershell was not found", function() end)
+        return
+      end
+      -- Should work with recommended config used in helper
+      helpers.set_shell_powershell()
+      eq('ああ\n', eval([[system('Write-Output "ああ"')]]))
+      -- Powershell expected to garble text without UTF8 console config
+      command([[let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command ']])
+      eq('??\n', eval([[system('Write-Output "ああ"')]]))
```
Comment (janlazo) [inline test/functional/eval/system_spec.lua:187]: Doesn't this depend on default input/output encoding? pwsh uses utf8 on linux.

Comment (dtcaciuc) [reply test/functional/eval/system_spec.lua]: Right, it does, should've tested in Linux env before pushing. Test now checks for different references for win and linux, PTAL

File: runtime/doc/options.txt (hunk 8 above)
Comment (dtcaciuc) [reply runtime/doc/options.txt]: Done, PTAL

Comment (dtcaciuc) [reply runtime/doc/options.txt]: Re: odd newline after the character, apparently that's what systemlist does on Windows? https://vim-jp.org/vimdoc-en/eval.html#systemlist()

Comment (janlazo) [reply runtime/doc/options.txt]: Yes, crlf is not converted to lf for systemlist.

**Review by janlazo — APPROVED** [review]
LGTM

File: test/functional/ex_cmds/make_spec.lua [hunk 11]
```diff
@@ -0,0 +1,45 @@
+local helpers = require('test.functional.helpers')(after_each)
+local clear = helpers.clear
+local eval = helpers.eval
+local has_powershell = helpers.has_powershell
+local has_pwsh = helpers.has_pwsh
```
Comment (janlazo) [inline test/functional/ex_cmds/make_spec.lua:5]: unused.

`make lint` failed.

Comment (dtcaciuc) [reply test/functional/ex_cmds/make_spec.lua]: Gah, sorry, fixed!
